rust/private/rust.bzl

```diff
@@ -1139,6 +1139,89 @@ rust_test_binary = rule(
         """),
 )
 
+def rust_test_suite(name, srcs, **kwargs):
+    """A rule for creating a test suite for a set of `rust_test` targets.
+
+    This rule can be used for setting up typical rust [integration tests][it]. Given the following
+    directory structure:
+
+    ```text
+    [crate]/
+        BUILD.bazel
+        src/
+            lib.rs
+            main.rs
+        tests/
+            integrated_test_a.rs
+            integrated_test_b.rs
+            integrated_test_c.rs
+            patterns/
+                fibonacci_test.rs
+    ```
+
+    The rule can be used to generate [rust_test](#rust_test) targets for each source file under `tests`
+    and a [test_suite][ts] which encapsulates all tests.
+
+    ```python
+    load("//rust:defs.bzl", "rust_binary", "rust_library", "rust_test_suite")
+
+    rust_library(
+        name = "math_lib",
+        srcs = ["src/lib.rs"],
+    )
+
+    rust_binary(
+        name = "math_bin",
+        srcs = ["src/main.rs"],
+    )
+
+    rust_test_suite(
+        name = "integrated_tests_suite",
+        srcs = glob(["tests/**"]),
+        deps = [":math_lib"],
+    )
+    ```
+
+    [it]: https://doc.rust-lang.org/rust-by-example/testing/integration_testing.html
+    [ts]: https://docs.bazel.build/versions/master/be/general.html#test_suite
+
+    Args:
+        name (str): The name of the `test_suite`.
+        srcs (list): All test sources, typically `glob(["tests/**/*.rs"])`.
+        **kwargs (dict): Additional keyword arguments. Args from [rust_test](#rust_test) are parsed out and shared
+            between all tests while [common attributes](https://docs.bazel.build/versions/master/be/common-definitions.html#common-attributes)
+            are shared between the test targets and `test_suite`.
+    """
+    tests = []
+
+    # Gather any rust attributes
+    rust_attrs = dict()
+    for attr in _common_attrs.keys():
+        if attr in kwargs:
+            value = kwargs.pop(attr)
+            if value:
+                rust_attrs.update({attr: value})
+
+    for attr in _rust_test_attrs.keys():
+        if attr in kwargs:
+            value = kwargs.pop(attr)
+            if value:
+                rust_attrs.update({attr: value})
+
+    for src in srcs:
+        test_name = src.replace(".rs", "").replace("/", "_")
```

Comment:
Is the `/`-stripping required, or just cosmetic? If the latter, let's maybe not do it?

Reply:
It's required, either this or we set an explicit `crate_name` but even then we'd be doing the same thing since `.` is not valid in a crate name.

Reply:
Does that end up changing how you'd refer to the individual test, from a bazel perspective?

for example setting the crate name explicitly would result in this being valid, right?
```bash
bazel test //test/rust_test_suite:tests/integrated_test_a
```

versus, this with the current version:
```bash
bazel test //test/rust_test_suite:tests_integrated_test_a
```

I would (slightly) lean toward the former.

Reply:
I don't think we can use the base name of the test. What happens in the case where there are two tests in different directories named the same?

Reply:
so bazel targets can have `/` characters in them, so I think you could leave them in.  I added a suggestion along the lines of what I was thinking.

Reply:
The targets now have forward slashes to match the directory structure.